src/libraries/Microsoft.PowerFx.Core/Binding/Binder.cs

```diff
@@ -3923,11 +3926,11 @@ public override void PostVisit(AsNode node)
                 _txb.SetInfo(node, new AsInfo(node, node.Right.Name));
 
                 var left = node.Left;
-                _txb.CheckAndMarkAsPageable(node);
+                bool isNodeStateful = _txb.CheckAndMarkAsPageable(node);
                 _txb.CheckAndMarkAsDelegatable(node);
                 _txb.SetType(node, _txb.GetType(left));
                 _txb.SetSideEffects(node, _txb.HasSideEffects(left));
-                _txb.SetStateful(node, _txb.IsStateful(left));
+                _txb.SetStateful(node, _txb.IsStateful(left) || isNodeStateful);
```

Comment:
When is an AsNode stateful but the left isn't?

Reply:
If this is valid, worth adding a comment explaining it

Reply:
In these few lines, we can set stateful flag twice.
Once in CheckAndMarkAsPageable and once on this line, in an inconsistent manner (true first time, false the second).

An expression like `Sum(Filter(t1, ThisRecord.Price > 50) As X, X.Price)` is when we hit this problem, with t1 being a Dataverse table.
On this code, `node` is `Filter(...) As X` and the left part is CallNode `Filter(...)`.

We start in CheckAndMarkAsPageable, the left node is pageable (line 1219) and as a result we set stateful to true at line 1227 on AsNode.
Then, on line 3933 we want to set stateful state on AsNode, based on the value of left Node (Filter CallNode) which is false.
We fail on Contracts.Assert at line 572.

Thinking of it, a much simpler approach is to simply remove line 3933!
The stateful flag is set by CheckAndMarkAsPageable when we need to and we don't need to do anything more here.